sound/soc/sof/intel/hda-mlink.c

```diff
@@ -14,28 +14,235 @@
+		h2link->elid = FIELD_GET(AZX_REG_ML_LEPTR_ID, h2link->leptr);
+
+		base_offset = FIELD_GET(AZX_REG_ML_LEPTR_PTR, h2link->leptr);
+		h2link->base_ptr = hlink->ml_addr + base_offset;
+
+		switch (h2link->elid) {
+		case AZX_REG_ML_LEPTR_ID_SDW:
+			h2link->shim_offset = AZX_REG_SDW_SHIM_OFFSET;
+			h2link->ip_offset = AZX_REG_SDW_IP_OFFSET;
+			h2link->shim_vs_offset = AZX_REG_SDW_VS_SHIM_OFFSET;
+			dev_dbg(dev, "Link %d: HDAudio extended - SoundWire alternate link, leptr.ptr %#x\n",
+				link_idx, base_offset);
+			break;
+		case AZX_REG_ML_LEPTR_ID_INTEL_DMIC:
+			h2link->shim_offset = AZX_REG_INTEL_DMIC_SHIM_OFFSET;
+			h2link->ip_offset = AZX_REG_INTEL_DMIC_IP_OFFSET;
+			h2link->shim_vs_offset = AZX_REG_INTEL_DMIC_VS_SHIM_OFFSET;
+			dev_dbg(dev, "Link %d: HDAudio extended - INTEL DMIC alternate link, leptr.ptr %#x\n",
+				link_idx, base_offset);
+			break;
+		case AZX_REG_ML_LEPTR_ID_INTEL_SSP:
+			h2link->shim_offset = AZX_REG_INTEL_SSP_SHIM_OFFSET;
+			h2link->ip_offset = AZX_REG_INTEL_SSP_IP_OFFSET;
+			h2link->shim_vs_offset = AZX_REG_INTEL_SSP_VS_SHIM_OFFSET;
+			dev_dbg(dev, "Link %d: HDAudio extended - INTEL SSP alternate link, leptr.ptr %#x\n",
+				link_idx, base_offset);
+			break;
+		case AZX_REG_ML_LEPTR_ID_INTEL_UAOL:
+			h2link->shim_offset = AZX_REG_INTEL_UAOL_SHIM_OFFSET;
+			h2link->ip_offset = AZX_REG_INTEL_UAOL_IP_OFFSET;
+			h2link->shim_vs_offset = AZX_REG_INTEL_UAOL_VS_SHIM_OFFSET;
+			dev_dbg(dev, "Link %d: HDAudio extended - INTEL UAOL alternate link, leptr.ptr %#x\n",
+				link_idx, base_offset);
+			break;
+		default:
+			dev_err(dev, "Link %d: HDAudio extended - Unsupported alternate link, leptr.id=%#02x value\n",
+				link_idx, h2link->elid);
+			return -EINVAL;
+		}
+	}
+
+	return 0;
+}
+
+/* END HDAML section */
+
+static int hda_ml_alloc_h2link(struct hdac_bus *bus, int i)
+{
+	struct hdac_ext2_link *h2link;
+	struct hdac_ext_link *hlink;
+	int ret;
+
+	h2link  = kzalloc(sizeof(*h2link), GFP_KERNEL);
+	if (!h2link)
+		return -ENOMEM;
+
+	/* basic initialization */
+	hlink = &h2link->hext_link;
+
+	hlink->index = i;
+	hlink->bus = bus;
+	hlink->ml_addr = bus->mlcap + AZX_ML_BASE + (AZX_ML_INTERVAL * i);
+
+	ret = hdaml_lnk_enum(bus->dev, h2link, hlink->ml_addr, i);
+	if (ret < 0) {
+		kfree(h2link);
+		return ret;
+	}
+
+	mutex_init(&h2link->eml_lock);
+
+	list_add_tail(&hlink->list, &bus->hlink_list);
+
+	/*
+	 * HDaudio regular links are powered-on by default, the
+	 * refcount needs to be initialized.
+	 */
+	if (!h2link->alt)
+		hlink->ref_count = 1;
+
+	return 0;
+}
+
 int hda_bus_ml_get_capabilities(struct hdac_bus *bus)
 {
-	if (bus->mlcap)
-		return snd_hdac_ext_bus_get_ml_capabilities(bus);
+	if (bus->mlcap) {
+		u32 link_count;
+		int ret;
+		int i;
+
+		link_count = readl(bus->mlcap + AZX_REG_ML_MLCD) + 1;
+
+		dev_dbg(bus->dev, "HDAudio Multi-Link count: %d\n", link_count);
+
+		for (i = 0; i < link_count; i++) {
+			ret = hda_ml_alloc_h2link(bus, i);
+			if (ret < 0) {
+				hda_bus_ml_free(bus);
```

Comment:
assuming h2link alloc failed with -ENOMEM, calling hda_bus_ml_free() would end up calling mutex_destroy(&h2link->eml_lock); but h2link would be NULL here isnt it?

Reply:
I don't think so, we only add the link to the list when there's no error, so on free that link would be skipped.

In other words, the hda_bus_ml_free() is to free what was allocated in the previous iterations.